src/services/api/errorHandling.ts

```diff
@@ -0,0 +1,81 @@
+import ky from 'ky';
+import {
+  ApiError,
+  NetworkError,
+  TimeoutError
+} from '~/services/api/httpErrors';
+import { JSONValue } from '~/services/api/types';
+
+export default async function handleError(e, setErrors) {
+  const { translatedError, errorMessage } = await mapError(e);
+
+  // invoke error hook with json answer (if any)
+  if (setErrors && errorMessage) {
+    setErrors(errorMessage);
+  }
+
+  throw translatedError;
+}
+
+/**
+ * Translate errors reported by ky/fetch to a set of custom exceptions
+ * which we can later on use to make decisions on how to handle specific errors.
+ */
+async function mapError(
+  e
+): Promise<{
+  errorMessage: string;
+  translatedError: any; // TODO: type Error
+}> {
+  let translatedError;
+  let errorMessage;
+  let statusCode;
+  switch (e.constructor) {
+    case ky.HTTPError: // a non 2xx error code was found
+      errorMessage = await parseErrorResponse(e.response);
+      statusCode = e.response.status;
+      translatedError = new ApiError(errorMessage, statusCode);
+      break;
+    case ky.TimeoutError:
+      translatedError = new TimeoutError(e.message);
+      break;
+    case TypeError:
+      translatedError = e; // throw as is
+      break;
+    default:
+      translatedError = new NetworkError(e); // pain point: every unexpected Error instance (that is not a TypeError) will be handled as NetworkError
+      break;
+  }
+  return { translatedError, errorMessage }; // return body content to not parse body multiple times, see https://github.com/node-fetch/node-fetch/issues/533
+}
```

Comment:
Now that I have had some time to think about this I am convinced that we should not translate ky errors at all. The handling of timeout and type errors adds nothing to the original error. For "network errors", the naming adds confusion because they could be something else than network errors entirely.

For HTTP errors, there is the problem that error messages may either exist as the text content of the response or nested in a JSON response body. That problem is not solved by wrapping the error thrown in a new error class.

Reply:
I thought about it some more and I like your approach now @s-pic